core/shared/src/main/scala/scalaz/zio/duration/Duration.scala

```diff
@@ -38,36 +38,30 @@ sealed trait Duration extends Ordered[Duration] with Serializable with Product {
 }
 
 final object Duration {
-
-  object Finite {
-
-    final def apply(nanos: Long): Duration =
-      if (nanos >= 0) new Finite(nanos)
-      else Infinity
-
-  }
-
   final case class Finite private (nanos: Long) extends Duration {
 
     final def +(other: Duration): Duration = other match {
-      case Finite(otherNanos) => Finite(nanos + otherNanos)
-      case Infinity           => Infinity
+      case Finite(otherNanos) =>
+        val sum = nanos + otherNanos
+        (nanos.signum, otherNanos.signum) match {
+          case (a, b) if a == b && a != sum.signum => Infinity // Integer overflow
+          case _                                   => Finite(sum)
+        }
+      case Infinity => Infinity
     }
 
     final def *(factor: Double): Duration =
-      if (!factor.isInfinite && !factor.isNaN) Finite((nanos * factor).round)
+      if (!factor.isInfinite && !factor.isNaN) Duration.fromNanos((nanos * factor).round)
```

Comment:
Missing check for overflow: `factor < Long.MaxValue / nanos` +- check for sign - with negative `nanos` it gets hairy but can be coded properly.